src/riak_kv_index_fsm.erl

```diff
@@ -194,30 +221,38 @@ process_results({Bucket, Results},
 process_results(done, StateData) ->
     {done, StateData}.
 
-finish({error, Error},
-       StateData=#state{from={raw, ReqId, ClientPid}}) ->
+finish({error, Error}, State=#state{from={raw, ReqId, ClientPid}}) ->
     %% Notify the requesting client that an error
     %% occurred or the timeout has elapsed.
     ClientPid ! {ReqId, {error, Error}},
-    {stop, normal, StateData};
+    {stop, normal, State};
 finish(clean,
-       StateData=#state{from={raw, ReqId, ClientPid}, merge_sort_buffer=undefined}) ->
+       State=#state{from={raw, ReqId, ClientPid},
+                    merge_sort_buffer=undefined}) ->
     ClientPid ! {ReqId, done},
-    {stop, normal, StateData};
+    log_timings(State#state.timings,
+                State#state.bucket,
+                State#state.results_sent),
+    {stop, normal, State};
 finish(clean,
        State=#state{from={raw, ReqId, ClientPid},
                     merge_sort_buffer=MergeSortBuffer,
                     results_sent=ResultsSent,
                     max_results=MaxResults}) ->
     LastResults = sms:done(MergeSortBuffer),
-    DownTheWire = case (ResultsSent + length(LastResults)) > MaxResults of
-                      true ->
-                          lists:sublist(LastResults, MaxResults - ResultsSent);
-                      false ->
-                          LastResults
-                  end,
+    TotalResults = length(LastResults) + ResultsSent,
+    DownTheWire =
+        case TotalResults > MaxResults of
```

Comment:
We don't need TotalResults. The function lists:sublist takes care of this, taking only the MaxResult - ResultSent is sufficient, making the code substantially easier:

Reply:
```suggestion
```

Reply:
MaxResults, it turns out can be the atom all as well as an integer(), and this relies on the fact any integer() > atom() is false.

So if we just have lists:sublist we get badarith.

But it is a bit worse, as it is also possible to construct a query whereby MaxResults is undefined, but a page-sort query is still requested.  undefined isn't expected by dialyzer.

Going to make the case of all vs integer() more specific, and ensure that not integer() is always all